### DISCUSSION TITLE
What happened to Homebrew?

### DISCUSSION THREAD

**Discussion by resistorsoftware** [Homebrew/discussions#1327] [Everyday usage]
Not sure when or what happened, but I was just reminded I could use a newer Ruby, 2.7.3. So I fired up rvm install 2.7.3 and that led to Homebrew being run, and it is doing a million things... none of which I asked for. A brief sample? 

    Installing gnupg dependency: libusb
    ==> Pouring libusb--1.0.24.catalina.bottle.tar.gz
    🍺  /usr/local/Cellar/libusb/1.0.24: 22 files, 516.7KB

Or this

     Upgrading gpgme 1.14.0 -> 1.15.1
    ==> Downloading https://ghcr.io/v2/homebrew/core/gpgme/manifests/1.15.1
    ######################################################################## 100.0%
    ==> Downloading https://ghcr.io/v2/homebrew/core/gpgme/blobs/sha256:d2f6e379b0c47d4b62275a588a69bd08fb1c79e153b348c96280f3162548a1e4
    ==> Downloading from https://pkg-containers-az.githubusercontent.com/ghcr1/blobs/sha256:d2f6e379b0c47d4b62275a588a69bd08fb1c79e153b348c96280f3162548a1e4?se=2021-04-21T01%3A45%3A00Z&sig=W10dK1AyZlU%2FOceM%2B8I
    ######################################################################## 100.0%

And on, and on, and on... what they heck is this? Crazy... So somewhere, somehow, my Homebrew is infected with crazy. What's the cure?

**Comment by SMillerDev** [comment]
> somehow, my Homebrew is infected with crazy. What's the cure?

Homebrew is updating all dependencies involved in something that `rvm` installed or updated from homebrew. It's simply making sure you have a working, up to date, system.

**Reply by resistorsoftware** [reply]
Yah.. so long ago there was PostgreSQL 9, then 10, then 11, then 12, now 13... and I use latest... 13... but this Homebrew session was updating 9, 10, 11, 12, 13... and this continues with all installed packages on my system. Just seems insane... I never asked Homebrew to do anything, it should not be launching into some massive update of sorts. It seems like some kind of update of "things it thinks it should know" rather than a big actual update, there was no compilation but at the same time, a TON of network activity and file changes I never asked for. 

Is this some kind of new behaviour of Homebrew? Does one turn this off?

**Reply by SMillerDev** [reply]
>  I never asked Homebrew to do anything, it should not be launching into some massive update of sorts.

You asked RVM to install something new, rvm depended on something new in Homebrew and Homebrew needed to update to fulfill this request.

>  there was no compilation but at the same time, a TON of network activity and file changes I never asked for.

Homebrew hasn't compiled software on the user machine since 2017 IIRC

> Is this some kind of new behaviour of Homebrew? Does one turn this off?

It is not, but it can seem a lot worse if you go long periods without updating.

---

This explains the reasoning for updates cascading:
https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-also-upgrade-a-bunch-of-other-stuff

**Comment by carlocab** [comment]
The cure is to regularly do `brew update && brew upgrade` so that you don't find yourself in a situation where `brew` discovers that it needs to upgrade a large number of formulae in order to install/upgrade something.

If you have specific version requirements for something you installed with `brew`, you should `brew extract` the formula to your own tap and install that version of the formula instead. (You can also try `brew pin`, but that can sometimes be fragile and cause more headaches in the long run than `brew extract`.)

**Reply by resistorsoftware** [reply]
The problem with Brew in my case was, I had a perfectly functioning dev machine. I went to update Ruby, and Homebrew interjected itself. In the end, that little Brew upgrade action tried to update my PostgreSQL even though my PostgreSQL was perfect and up to date. So it died finding some old files around, and stiffed me with a non-working PostgreSQL that could not be easily fixed because of version incompatibilities. The hard work would've been wasting hours trying to resurrect that. The easy work was just blowing it all away and starting over. 

Again, I am not naive, but this Homebrew has evolved into something beyond a helper for those too lazy to compile from source. It is now a "hard to use", "difficult to maintain", "savage ruiner of good things"... so my opinion matter zero, I contributed zero to this project, I am just sad it went so poorly for me, eventually leading me to abandoning it before it causes me anymore grief.

**Reply by carlocab** [reply]
You probably had the `postgresql` formula installed, which offers no guarantees about the version.

Both `ruby` and `postgresql` depend on `openssl@1.1`. It's likely that your `openssl@1.1` install was out-of-date, so upgrading `ruby` would also upgrade `openssl@1.1`, and therefore also `postgresql` (and anything else out-of-date that is connected to `ruby` via the dependency graph).

**Reply by resistorsoftware** [reply]
There is no question on any of that. That being said, the fact I ended up with a broken system is what ticks me off. I don't mind breaking things myself. I have been compiling PostgreSQL and setting it up since version 8. But yesterday, Homebrew set me back and pretty much destroyed my working system for no really good reason. Sure, maybe OpenSSL did need an upgrade. No excuse for Homebrew going through my system, deciding PostgreSQL needed updating, and then managing to not only work on the wrong version, but to also put it into a state where manual rescue of the damage was almost impossible.

**Reply by carlocab** [reply]
Which PostgreSQL formula did you have installed originally? Which version of PostgreSQL did you have installed, and which version did `brew` try to upgrade it to?

**Reply by resistorsoftware** [reply]
No idea. All I know is my /usr/local/var had /postgres, postgres@9, postgres@10, postgres@11 and a postgres@12 all the while I was using PostgreSQL 13 for my day to day work... So yes, you could say there was a lot of junk there... but still. So the Homebrew update got caught at version 12... and could not update to version 13 because version 13 was too far ahead. 

The reality was, it should have recognized the current use of version 13, thus compiling that with the new OpenSSL and then be done. But no... it went beserkers. Sorry for not housekeeping better. I never honestly thought an upgrade would get so lost.

**Reply by carlocab** [reply]
Sorry, but this really makes it impossible to tell if something in `brew` is broken or if it's something in your local config. For example, I have no idea what "it went beserkers" means here. Nor is it clear what "it got caught at version 12" or "could not update to version 13 because version 13 was too far ahead" means.

The exact output from `brew` here would've been useful in order to clarify what `brew` was doing (and hopefully ascertain what went wrong).

I haven't encountered any similar bug reports (the only ones that come to mind are ones about difficulties migrating between major versions of PostgreSQL), so it seems likely to me that the problem is in your local config. I think the fact that you're not even sure which PostgreSQL formula you had installed suggests that you may very well have unknowingly done something to your setup that made it susceptible to breakage with `brew upgrade`.

**Reply by resistorsoftware** [reply]
I was live and booting PostgreSQL 13. For some reason, brew noticed postgresql@12 and so it took that as an opportunity to update to 13. So even though I was on 13 already, it took 12 and tried to go to 13, which broke, as it could not. I went through all the logs and discovered this. So it left my current setup in a shambles. You try fixing that. 

As for what I meant by brew went beserkers, my screen just kept filling with formula after formula all being copied, downloaded and used for something. What really drove me nuts was Python too. I never use Python for anything, but it went ahead and worked on some versions that are not even valid anymore. 

In all those logs too, a ton of notices about not being able to link thinks properly, so happy danger in the future... 

Anyway, call me the problem. I am not calling brew the problem per se. But I am saying a utility like brew is now something I think I need to remove from life, as it is "TOO BIG SO THAT WHEN IT FAILS, it fails bad". I don't need the convenience if it comes with the risk of it breaking my setup. 

In other words, I could care a less if it happily reminds me to update shit that needs updating for all intents, and if it can carefully and accurately do so. But if it is just an atom bomb, that goes off on a mission to do too much, then it sucks, as I would prefer to keep my system working and updated than broken and who knows. 

I do not blame brew or anyone for this, I guess I am just pointing out, brew has outlived it's usefulness to me, it is now more hassle to use it than it is worth.

**Reply by resistorsoftware** [reply]
PS: You are correct in that most of this problem, in fact all of it, stems from PostgreSQL and the scripts assigned to it to update versions. The sad thing is, I was already using the latest @13 and needed no help there at all, other than a recompile if indeed OpenSSL was in need of an update.

How on earth it got bogged down in my history of having versions 9, 10, 11 and 12 versions sitting around doing nothing is a hard question for that. They were there only as a history, and as a "whoa, what if one day I need this version again?"

**Reply by carlocab** [reply]
[`postgresql@12`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/postgresql@12.rb) is PostgreSQL 12.6, so if you had that installed, then `brew` shouldn't, and wouldn't, have upgraded that to 13.

Python depends on `openssl@1.1` and is therefore connected to `ruby` via the dependency graph. See

>and anything else out-of-date that is connected to `ruby` via the dependency graph

Indeed, very many formulae depend on `openssl@1.1`, so having to upgrade OpenSSL will often force many other installed formulae to be on their latest version. That's not a problem if you regularly do `brew update && brew upgrade`, because this keeps things you have installed at their latest version, so you never really encounter `brew` "going beserkers". If you wait long periods between doing this, then I'm not sure it should be surprising to find that `brew` finds many things outdated that it needs to upgrade before you can install/upgrade something.

>I do not blame brew or anyone for this, I guess I am just pointing out, brew has outlived it's usefulness to me, it is now more hassle to use it than it is worth.

Ok. Good to know, I guess.

**Comment by tlk** [comment]
> my Homebrew is infected with crazy. What's the cure?

Sorry to hear about that, I also have had some .. disappointing experiences. I'm afraid it may have been working as intended, though obviously not as you expected. I don't know anything about rvm or the many versions of postgresql so I may be off in the following.

If you are used to Debian/Ubuntu `apt-get`, *BSD ports or other package managers, Homebrew works slightly different than those. It used to be such that you would have to run `brew update` manually, and installing packages in between would not trigger an avalanche of package upgrades, but five(?) years ago it was changed such that it runs `brew  update` automatically when you run `brew install foobar`.

In principle this is a good thing because it keeps your system updated, but it can also trigger an avalanche of cascading upgrades at an inconvenient time. This default setting can be changed so you have to run `brew update` manually. This is how I prefer to use Homebrew, but please note that it is not recommended by the Homebrew maintainers - and importantly, one has to remember to run `brew update` every now and then.

You may find the discussion in https://github.com/Homebrew/discussions/discussions/1107 useful.

**Reply by tlk** [reply]
Just wanted to add two commands that seems to be essential for a happy Homebrew.

This is needed surprisingly often:
```
xcode-select --install
```

This one will uninstall formulae that were only installed as a dependency of another formula and are now no longer needed. Perhaps this would have gotten rid of the old postgresql versions on your system.

```
brew autoremove
```

**Comment by resistorsoftware** [comment]
Thank you everyone for your kind comments and helpful insights. Much appreciated.
